<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>

## Introduction

As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the keras library. 

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3> 
1. <a href="#item1">Download and Clean Dataset</a>  
2. <a href="#item2">Import Keras</a>  
3. <a href="#item3">Build a Neural Network</a>  
4. <a href="#item4">Train and Test the Network</a>  
</font>
</div>

<a id="item1"></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [46]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different materials that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [47]:
concrete_data = pd.read_csv('https://ibm.box.com/shared/static/svl8tu7cmod6tizo6rk0ke4sbuhtpdfx.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [48]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [49]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [50]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

In [51]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>

Let's do a quick sanity check of the predictors and the target dataframes.

In [52]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [53]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by subtracting the mean and dividing by the standard deviation.

In [54]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [55]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

## Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

#### Let's go ahead and import the Keras library

In [56]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [57]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [58]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(500, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(350, activation='relu'))
    model.add(Dense(200, activation='relu'))
    model.add(Dense(100, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

<a id="item4"></a>

## Train and Test the Network

In [59]:
# build the model
rmodel = regression_model()

# fit the model
rmodel.fit(predictors_norm, target, validation_split=0.3, epochs=200, verbose=2)
print('Done')

Train on 721 samples, validate on 309 samples
Epoch 1/200
 - 0s - loss: 915.9974 - val_loss: 251.3249
Epoch 2/200
 - 0s - loss: 250.7284 - val_loss: 216.3738
Epoch 3/200
 - 0s - loss: 196.2144 - val_loss: 214.7914
Epoch 4/200
 - 0s - loss: 173.7758 - val_loss: 202.1399
Epoch 5/200
 - 0s - loss: 152.2963 - val_loss: 199.3926
Epoch 6/200
 - 0s - loss: 132.4068 - val_loss: 201.3561
Epoch 7/200
 - 0s - loss: 111.8033 - val_loss: 213.0029
Epoch 8/200
 - 0s - loss: 91.3457 - val_loss: 209.5833
Epoch 9/200
 - 0s - loss: 76.5771 - val_loss: 193.9053
Epoch 10/200
 - 0s - loss: 64.4710 - val_loss: 191.5299
Epoch 11/200
 - 0s - loss: 54.7349 - val_loss: 180.8996
Epoch 12/200
 - 0s - loss: 51.4933 - val_loss: 167.7364
Epoch 13/200
 - 0s - loss: 55.7398 - val_loss: 163.3329
Epoch 14/200
 - 0s - loss: 44.7874 - val_loss: 185.3980
Epoch 15/200
 - 0s - loss: 38.3732 - val_loss: 179.4522
Epoch 16/200
 - 0s - loss: 35.2756 - val_loss: 143.6867
Epoch 17/200
 - 0s - loss: 33.3936 - val_loss: 124.5101
Epoc

Epoch 147/200
 - 0s - loss: 9.1749 - val_loss: 175.7819
Epoch 148/200
 - 0s - loss: 8.6757 - val_loss: 121.0570
Epoch 149/200
 - 0s - loss: 8.4833 - val_loss: 114.2169
Epoch 150/200
 - 0s - loss: 8.7564 - val_loss: 126.3785
Epoch 151/200
 - 0s - loss: 7.6971 - val_loss: 148.4249
Epoch 152/200
 - 0s - loss: 8.9081 - val_loss: 120.6513
Epoch 153/200
 - 0s - loss: 7.7830 - val_loss: 150.5120
Epoch 154/200
 - 0s - loss: 7.5112 - val_loss: 128.8988
Epoch 155/200
 - 0s - loss: 8.9197 - val_loss: 206.2534
Epoch 156/200
 - 0s - loss: 8.7959 - val_loss: 135.9927
Epoch 157/200
 - 0s - loss: 7.7635 - val_loss: 134.0134
Epoch 158/200
 - 0s - loss: 7.1528 - val_loss: 130.7212
Epoch 159/200
 - 0s - loss: 6.6873 - val_loss: 128.2081
Epoch 160/200
 - 0s - loss: 7.1240 - val_loss: 140.8057
Epoch 161/200
 - 0s - loss: 8.0737 - val_loss: 164.7652
Epoch 162/200
 - 0s - loss: 9.6445 - val_loss: 111.2993
Epoch 163/200
 - 0s - loss: 9.1579 - val_loss: 169.7569
Epoch 164/200
 - 0s - loss: 7.0787 - val_loss: 1

<strong>You can refer to this [link](https://keras.io/models/sequential/) to learn about other functions that you can use for prediction or evaluation.</strong>

In [60]:
sample_data = concrete_data.sample(5)
sample_predictors = sample_data[concrete_data_columns[concrete_data_columns != 'Strength']]
sample_predictors_norm = (sample_predictors - predictors.mean()) / predictors.std()
print(sample_data.head())
rmodel.predict(sample_predictors_norm, verbose=1)


     Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
161   337.9               189.0      0.0  174.9               9.5   
848   252.0                97.0     76.0  194.0               8.0   
601   339.0                 0.0      0.0  197.0               0.0   
178   286.3               200.9      0.0  144.7              11.2   
269   231.8                 0.0    121.6  174.0               6.7   

     Coarse Aggregate  Fine Aggregate  Age  Strength  
161             944.7           755.8   56     59.89  
848             835.0           821.0   28     33.40  
601             968.0           781.0   28     32.04  
178            1004.6           803.7   91     76.80  
269            1056.4           778.5    3     15.44  
5/5 [==============================] - 0s 11ms/step


array([[59.803814],
       [35.54224 ],
       [31.934269],
       [76.19837 ],
       [16.272955]], dtype=float32)

Feel free to vary the following and note what impact each change has on the model's performance:

1. Increase or decrease number of neurons in hidden layers
2. Add more hidden layers
3. Increase number of epochs

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!

This notebook is part of a course on **edX** called *Deep Learning Fundamentals with Keras*. If you accessed this notebook outside the course, you can take this course online by clicking [here](http://cocl.us/DL0101EN_edX_Week3_LAB1).

<hr>

Copyright &copy; 2018 [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).